### ISSUE TITLE
Writing to buckets located in Europe

### ISSUE THREAD

**Issue by StanBright** [marcel/aws-s3#4]
Currently there's problem with writing to buckets that are located in Europe. That's because we need the bucket name as part of the domain name, rather than the path itself.

We have a fork of this library that supports writing to buckets. Is it possible to have a merge of those changes?

http://github.com/isaacfeliu/aws-s3/tree/master

**Comment by samlown** [comment]
Alternatively this commit is a lot more lightweight and has tests:

http://github.com/fesplugas/aws-s3/commit/90c30e79dc2327776308ba07c207a2a872325a3c

**Comment by donaldpiret** [comment]
Wondering if this issue is ever going to get resolved, it's a pity as aws-s3 is such a great library and it's being used by Paperclip too, but no european bucket support is a pain for many many users

**Comment by samlown** [comment]
I've now converted to "carrierwave" and right_aws instead of paperclip. Aws-s3 is designed for the old non-euro Amazon API so all modifications to add Euro support look like nasty hacks and are difficult to test. RightAWS however uses the new API, if you avoid the helper Bucket/Key models and use the S3Interface directly it avoids many remote calls and performs very well. cheers, sam.

**Comment by StanBright** [comment]
Well, this lib - isaacfeliu/aws-s3 is working good for our project.

**Comment by samlown** [comment]
I don't think this is just a question of "using the bucket as part of the domain name" and resolving that issue with a patch. The S3 API has changed (a long time ago) and aws-s3 hasn't kept up. Also, Amazon have just lowered their pricing on EU buckets, meaning this will be even more popular, and it wouldn't surprise me if they start launching data centers in Asia soon.

Use an alternative that works, as I'm certainly not going to oblige Marcel to fix this. Sam

**Comment by jhchabran** [comment]
@samlown : using an alternative is fine you're right, but it's a bit messy to manage vanilla gems and forked ones. At least could you write something about EU buckets in the README to avoid people to loose time like I did ?

**Comment by greatseth** [comment]
The patch mentioned here looks pretty clean to me, and it seems to work great. Thanks to all responsible! Would sure be nice to get this merged into the proper repo, though..

**Comment by tlray** [comment]
Any progress?

**Comment by tlray** [comment]
Thanks, that looks great!

**Comment by fjfish** [comment]
Change config/environment.rb

Add this at the bottom

AWS::S3::DEFAULT_HOST = "s3-eu-west-1.amazonaws.com"

You will see a warning about reinitializing the constant, but it works fine.

**Comment by fjfish** [comment]
Of course, AWS::S3.set_default_host, would be better ;)

**Comment by elia** [comment]
Using replace you can avoid those ugly warnings...
    AWS::S3::DEFAULT_HOST.replace "s3-eu-west-1.amazonaws.com"

**Comment by xwmx** [comment]
Note that this isn't just an issue with European buckets. The same issue exists with US-West and, presumably, Singapore buckets.

**Comment by elia** [comment]
Also this works

```
class << AWS::S3
  ENDPOINTS = {
    :us     => 's3-us-west-1.amazonaws.com', 
    :asia   => 's3-ap-southeast-1.amazonaws.com', 
    :europe => 's3-eu-west-1.amazonaws.com'
  }

  def Base.region= region
    endpoint = ENDPOINTS[region] or 
      raise S3Exception, "Unknown region: please use one of #{ENDPOINTS.keys.map(&:inspect).join(', ')}."
    DEFAULT_HOST.replace endpoint
  end
end

AWS::S3::Base.region = :europe
```

it should be included by our beloved author, marcel...

**Comment by parndt** [comment]
I can't believe this is still an open issue when there are so many solutions offered and aws-s3 is the standard way to use s3 buckets. This is a common issue I hear/see raised regarding s3. :(

**Comment by fjfish** [comment]
Maybe Marcel thinks we can just use these solutions while he does his day job? Not ideal, but I'm very grateful he did the majority of the work for me. :)

**Comment by goz** [comment]
elia : it did work for upload, while paperclip still showing s3.amazonaws.com URL, how do you solved it?

**Comment by elia** [comment]
@gozali: I ended up using the [s3 gem](http://rubygems.org/gems/s3)

**Comment by murdoch** [comment]
I got paperclip working with s3 and eu-buckets after reading this tutorial: http://www.conandalton.net/2011/02/paperclip-s3-and-european-buckets.html

**Comment by parndt** [comment]
I switched to fog. :)

**Comment by manuelmeurer** [comment]
+1 for the solution on conandalton.net. Worked like a charm with one little fix, see comments.

**Comment by batizhevsky** [comment]
@ella solution is work for me.
Why this is problem is steel not solved?

**Comment by elia** [comment]
@leonko I think you should ask the author

**Comment by batizhevsky** [comment]
@marcel  can you add this changes in trunk, please?

**Comment by manuelmeurer** [comment]
@leonko It's quite unfriendly and inappropriate to "yell" at people like that. @marcel is in no way obligated to fix this for you. I realize it is not an ideal situation that this issue has not yet been resolved, but you can always fork the project, do whatever modifications you need and use your forked version.

**Comment by cyrildiagne** [comment]
wow just had a hard time debugging a very simple sinatra file upload using amazon s3 because of this issue! created a US standard bucket, and it now works like a charm!
